File: algorithm_data_structure/leetcode/144.binary-tree-preorder-traversal.py
#
# @lc app=leetcode id=144 lang=python3
#
# [144] Binary Tree Preorder Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # Use a stack and two color approach! Miao A!
        WHITE, BLACK = 0, 1
        res = []
        stack = []
        stack.append((WHITE, root))

        while stack:
            color, node = stack.pop()
            if node is None:
                continue
            if color == WHITE:
                stack.append((WHITE, node.right))
                stack.append((WHITE, node.left))
                stack.append((BLACK, node))  # Last in first out!
            else:
                res.append(node.val)

        return res

        # Use recursion
    #     if not root:
    #         return []

    #     self.res = []
    #     self.helper(root)
    #     return self.res

    # def helper(self, node):
    #     self.res.append(node.val)
    #     if node.left:
    #         self.helper(node.left)
    #     if node.right:
    #         self.helper(node.right)


# @lc code=end

File: algorithm_data_structure/leetcode/144.binary-tree-preorder-traversal.cpp
/*
 * @lc app=leetcode id=144 lang=cpp
 *
 * [144] Binary Tree Preorder Traversal
 */

// @lc code=start

#include <algorithm>
#include <stack>
#include <vector>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
   public:
	vector<int> preorderTraversal(TreeNode *root) {
		vector<int> ans;
		if (not root) {
			return ans;
		}

		stack<TreeNode *> s;
		s.push(root);

		while (not s.empty()) {
			TreeNode *node = s.top();
			s.pop();
			ans.push_back(node->val);

			// Firstly push the right side in the queue, so the left node is
			// firstly poped out from the stack
			if (node->right) {
				s.push(node->right);
			}

			if (node->left) {
				s.push(node->left);
			}
		}

		return ans;
	}
};
// @lc code=end
